tinkerpop3/src/main/java/io/shiftleft/overflowdb/storage/OdbStorage.java

```diff
@@ -98,13 +101,70 @@ public NodeSerializer getNodeSerializer() {
 
   public MVMap<Long, byte[]> getNodesMVMap() {
     if (mvstore == null) {
-      mvstore = new MVStore.Builder().fileName(mvstoreFile.getAbsolutePath()).open();
-      nodesMVMap = mvstore.openMap("nodes");
+      mvstore = initializeMVStore();
     }
+    if (nodesMVMap == null)
+      nodesMVMap = mvstore.openMap("nodes");
     return nodesMVMap;
   }
 
+  private MVStore initializeMVStore() {
+    final MVStore store = new MVStore.Builder().fileName(mvstoreFile.getAbsolutePath()).open();
+    return store;
+  }
+
+  private Map<String, String> getIndexNameMap(MVStore store) {
+    store.getMapNames().stream().forEach(name -> System.out.println("name: " + name));
+    return store
+        .getMapNames()
+        .stream()
+        .filter(s -> s.startsWith(INDEX_PREFIX))
+        .map(s -> {System.out.println("filtered name: " + s); return s;})
+        .collect(Collectors.toConcurrentMap(s -> removeIndexPrefix(s), s -> s));
+  }
+
+  public Set<String> getIndexNames() {
+    return getIndexNameMap(mvstore).keySet();
+  }
+
+  private String removeIndexPrefix(String s) {
+    assert s.startsWith(INDEX_PREFIX);
+    return s.substring(INDEX_PREFIX.length());
+  }
+
+  public MVMap<Object, long[]> openIndex(String indexName) {
+    System.out.println("Opening index: " + indexName);
+    assert indexName != null && !indexName.isEmpty();
+    final String mapName = getIndexMapName(indexName);
+    System.out.println("Opening index nvmap: " + mapName);
+    MVMap<Object, long[]> indexMVMap = mvstore.openMap(mapName);
+    return indexMVMap;
+  }
+
+  private String getIndexMapName(String indexName) {
+    return INDEX_PREFIX + indexName;
+  }
+
+
   public Optional<NodeDeserializer> getNodeDeserializer() {
     return nodeDeserializer;
   }
+
+  public void clearIndices() {
+    getIndexNames().forEach(this::clearIndex);
+  }
+
+  private void clearIndex(String indexName) {
+    System.out.println("clearing index: " + indexName);
+    openIndex(indexName).clear();
+  }
+
+  public void saveIndex(String propertyName, Map<Object, Set<NodeRef>> indexMap) {
+    final MVMap<Object, long[]> indexStore = openIndex(propertyName);
```

Comment:
worth checking: are there any limitations on mvmap names, so that we'd need to adjust/encode the names?

Reply:
I've checked the code and it does not look that there are any limitations